File: trunk/phetcommon/src/edu/colorado/phet/common/math/Vector2D.java
/**
 * Class: ImmutableVector2D
 * Package: edu.colorado.phet.common.math
 * Author: Another Guy
 * Date: May 20, 2004
 */
package edu.colorado.phet.common.math;

public interface Vector2D extends ImmutableVector2D {
    Vector2D add( ImmutableVector2D v );

    Vector2D scale( double scale );

    Vector2D subtract( ImmutableVector2D that );

    void setX( double x );

    void setY( double y );

    void setCoordinates( double x, double y );

    Vector2D normalize();

    public static class Double extends ImmutableVector2D.Double implements Vector2D {
        public Double() {
        }

        public Double( double x, double y ) {
            super( x, y );
        }

        public Vector2D add( ImmutableVector2D v ) {
            setX( getX() + v.getX() );
            setY( getY() + v.getY() );
            return this;
        }

        public Vector2D normalize() {
            double length = getMagnitude();
            return scale( 1.0 / length );
        }

        public Vector2D scale( double scale ) {
            setX( getX() * scale );
            setY( getY() * scale );
            return this;
        }

        public void setX( double x ) {
            super.setX( x );
        }

        public void setY( double y ) {
            super.setY( y );
        }

        public void setCoordinates( double x, double y ) {
            setX( x );
            setY( y );
        }

        public Vector2D subtract( ImmutableVector2D that ) {
            setX( getX() - that.getX() );
            setY( getY() - that.getY() );
            return this;
        }

        public String toString() {
            return "Vector2D.Double[" + getX() + ", " + getY() + "]";
        }
    }

}

File: simulations-java/simulations/moving-man/scala-src/edu/colorado/phet/movingman/ladybug/Vector2D.scala
package edu.colorado.phet.movingman.ladybug

class Vector2D(_x: Double, _y: Double) {
  val x = _x
  val y = _y

  override def toString = "x=" + x + ", y=" + y

  def this() = this (0, 0)

  def +(that: Vector2D) = new Vector2D(that.x + x, that.y + y)

  def getAngle = java.lang.Math.atan2(y, x)
}

object TestVector2D {
  def main(args: Array[String]) = {
    val a = new Vector2D(3, 2)
    val b = new Vector2D(5, 5)
    println(a)
    println(a + b)
  }
}